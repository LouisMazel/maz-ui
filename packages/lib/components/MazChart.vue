<template>
  <Component :is="component" v-bind="propsRef" />
</template>

<script lang="ts" setup>
  import { defineAsyncComponent } from 'vue'
  import { type ChartProps } from 'vue-chartjs'
  import {
    ArcElement,
    BarElement,
    LineElement,
    PointElement,
    CategoryScale,
    Chart,
    Legend,
    LinearScale,
    Title,
    Tooltip,
  } from 'chart.js'
  import { ref, type PropType } from 'vue'

  const props = defineProps({
    /**
     * Chart.js chart type
     */
    type: { type: String as PropType<ChartProps['type']>, required: true },
    /**
     * The data object that is passed into the Chart.js chart
     * @see https://www.chartjs.org/docs/latest/getting-started/
     */
    data: { type: Object as PropType<ChartProps['data']>, required: true },
    /**
     * The options object that is passed into the Chart.js chart
     * @see https://www.chartjs.org/docs/latest/general/options.html
     */
    options: { type: Object as PropType<ChartProps['options']>, default: Object },
    /**
     * The plugins array that is passed into the Chart.js chart
     * @see https://www.chartjs.org/docs/latest/developers/plugins.html
     */
    plugins: { type: Array as PropType<ChartProps['plugins']>, default: Array },
    /**
     * Key name to identificate dataset
     */
    datasetIdKey: { type: String as PropType<ChartProps['datasetIdKey']>, default: 'label' },
    /**
     * A mode string to indicate transition configuration should be used.
     * @see https://www.chartjs.org/docs/latest/developers/api.html#update-mode
     */
    updateMode: { type: String as PropType<ChartProps['updateMode']>, default: undefined },
  })

  const propsRef = ref<unknown>(props)

  Chart.register(
    CategoryScale,
    LinearScale,
    Title,
    Tooltip,
    Legend,
    BarElement,
    ArcElement,
    PointElement,
    LineElement,
  )

  const component = defineAsyncComponent(async () => {
    const { Bar, Bubble, Doughnut, Line, Pie, PolarArea, Radar, Scatter } = await import(
      'vue-chartjs'
    )

    switch (props.type) {
      case 'bar': {
        return Bar
      }
      case 'line': {
        return Line
      }
      case 'scatter': {
        return Scatter
      }
      case 'bubble': {
        return Bubble
      }
      case 'pie': {
        return Pie
      }
      case 'doughnut': {
        return Doughnut
      }
      case 'polarArea': {
        return PolarArea
      }
      case 'radar': {
        return Radar
      }
    }
  })
</script>
