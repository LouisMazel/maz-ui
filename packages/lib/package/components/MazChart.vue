<script lang="ts">
  import { h, ref, onMounted, defineComponent, PropType } from 'vue'

  import {
    chartJsEventNames,
    generateEventObject,
    generateChartJsEventListener,
  } from './MazChart/includes'

  import {
    Chart,
    ArcElement,
    LineElement,
    BarElement,
    PointElement,
    BarController,
    BubbleController,
    DoughnutController,
    LineController,
    PieController,
    PolarAreaController,
    RadarController,
    ScatterController,
    CategoryScale,
    LinearScale,
    LogarithmicScale,
    RadialLinearScale,
    TimeScale,
    TimeSeriesScale,
    Decimation,
    Filler,
    Legend,
    Title,
    Tooltip,
    ChartItem,
    ChartType,
    Plugin,
    ChartData,
    ChartOptions,
  } from 'chart.js'

  Chart.register(
    ArcElement,
    LineElement,
    BarElement,
    PointElement,
    BarController,
    BubbleController,
    DoughnutController,
    LineController,
    PieController,
    PolarAreaController,
    RadarController,
    ScatterController,
    CategoryScale,
    LinearScale,
    LogarithmicScale,
    RadialLinearScale,
    TimeScale,
    TimeSeriesScale,
    Decimation,
    Filler,
    Legend,
    Title,
    Tooltip,
  )

  export default defineComponent({
    name: 'MazChart',
    props: {
      type: { type: String, required: true },
      data: { type: Object, required: true },
      height: { type: String, default: undefined },
      width: { type: String, default: undefined },
      options: { type: Object, default: Object },
      plugins: { type: Array as PropType<Plugin[]>, default: () => [] },
      locale: { type: String, default: 'fr-FR' },
    },
    emits: chartJsEventNames,
    setup(props, { emit }) {
      const chartRef = ref<HTMLCanvasElement>()
      //generate chart.js plugin to emit lib events
      const chartJsEventsPlugin: Plugin = chartJsEventNames.reduce(
        (reduced, eventType) => {
          const event = generateEventObject(eventType, chartRef.value)
          return { ...reduced, ...generateChartJsEventListener(emit, event) }
        },
        { id: 'MazChartEventHookPlugin' },
      )

      type ChartJssState = {
        chart?: Chart
        options?: ChartOptions
        plugins: Plugin[]
        props: typeof props
      }

      const chartJSState: ChartJssState = {
        chart: undefined,
        plugins: [chartJsEventsPlugin, ...props.plugins],
        props: { ...props },
      }

      const destroy = () => {
        if (chartJSState.chart) {
          chartJSState.chart.destroy()
          chartJSState.chart = undefined
        }
      }

      const update = () => {
        if (chartJSState.chart) {
          chartJSState.chart.data = {
            ...chartJSState.chart.data,
            ...chartJSState.props.data,
          }
          chartJSState.chart.options = {
            ...chartJSState.chart.options,
            ...chartJSState.props.options,
          }
          chartJSState.chart.update('resize')
        }
      }

      const resize = () => chartJSState.chart && chartJSState.chart.resize()

      const render = () => {
        if (chartJSState.chart) {
          return chartJSState.chart.update()
        }
        if (chartRef.value) {
          return (chartJSState.chart = new Chart(
            chartRef.value.getContext('2d') as ChartItem,
            {
              type: chartJSState.props.type as ChartType,
              data: chartJSState.props.data as ChartData,
              options: {
                locale: props.locale,
                ...chartJSState.props.options,
              },
              plugins: chartJSState.plugins,
            },
          ))
        }
      }

      onMounted(() => render())

      return {
        chartJSState,
        chartRef,
        render,
        resize,
        update,
        destroy,
      }
    },
    render(props) {
      return h('canvas', {
        ref: 'chartRef',
        width: props.width,
        height: props.height,
      })
    },
  })
</script>
