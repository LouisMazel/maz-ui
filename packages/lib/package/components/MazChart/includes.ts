export const chartJsEventNames = [
  'install',
  'start',
  'stop',
  'beforeInit',
  'afterInit',
  'beforeUpdate',
  'afterUpdate',
  'beforeElementsUpdate',
  'reset',
  'beforeDatasetsUpdate',
  'afterDatasetsUpdate',
  'beforeDatasetUpdate',
  'afterDatasetUpdate',
  'beforeLayout',
  'afterLayout',
  'afterLayout',
  'beforeRender',
  'afterRender',
  'resize',
  'destroy',
  'uninstall',
  'afterTooltipDraw',
  'beforeTooltipDraw',
]

export function generateEventObject(
  type: string,
  chartRef?: HTMLCanvasElement,
) {
  //chart js allows some events to be cancelled if they return false
  //this implements familiar logic to allow vue emitted chart events to be canceled
  return {
    type: type,
    chartRef: chartRef,
    preventDefault() {
      this._defaultPrevented = true
    },
    isDefaultPrevented() {
      return !this._defaultPrevented
    },
    _defaultPrevented: false,
  }
}

export function generateChartJsEventListener(
  emit: (event: string, ...args: any[]) => void,
  event: ReturnType<typeof generateEventObject>,
) {
  return {
    [event.type]: () => {
      emit(event.type, event)
      return event.isDefaultPrevented()
    },
  }
}
