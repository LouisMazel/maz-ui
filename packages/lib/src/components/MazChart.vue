<script lang="ts">
/* eslint-disable import/first */
import type {
  ChartData,
  ChartOptions,
  ChartType,
  DefaultDataPoint,
  Plugin,
  UpdateMode,
} from 'chart.js'

export type { ChartData as MazChartData, DefaultDataPoint as MazChartDefaultDataPoint, Plugin as MazChartPlugin, ChartType as MazChartType, UpdateMode as MazChartUpdateMode } from 'chart.js'

export interface MazChartProps<T extends ChartType = ChartType, TData = DefaultDataPoint<T>, TLabel = unknown> {
  /**
   * Type of the chart
   * @type ChartType
   * @values 'bar', 'line', 'scatter', 'bubble', 'pie', 'doughnut', 'polarArea', 'radar'
   * @required
   */
  type: T
  /**
   * Data of the chart
   * @type ChartData<T, TData, TLabel>
   * @required
   */
  data: ChartData<T, TData, TLabel>
  /**
   * Options of the chart
   * @type ChartOptions<T>
   * @default {}
   */
  options?: ChartOptions<T>
  /**
   * Plugins of the chart
   * @type Plugin[]
   */
  plugins?: Plugin[]
  /**
   * Dataset ID key
   * @type string
   */
  datasetIdKey?: string
  /**
   * Update mode
   * @type UpdateMode
   * @values 'resize', 'reset', 'none', 'hide', 'show', 'default', 'active'
   * @default 'default'
   */
  updateMode?: UpdateMode
}
</script>

<script lang="ts" setup generic="T extends ChartType, TData = DefaultDataPoint<T>, TLabel = unknown">
import {
  ArcElement,
  BarElement,
  CategoryScale,
  Chart,
  Legend,
  LinearScale,
  LineElement,
  PointElement,
  Title,
  Tooltip,
} from 'chart.js'
import { defineAsyncComponent } from 'vue'

const {
  type,
  data,
  options = {},
  plugins,
  datasetIdKey,
  updateMode,
} = defineProps<MazChartProps<T, TData, TLabel>>()

Chart.register(
  CategoryScale,
  LinearScale,
  Title,
  Tooltip,
  Legend,
  BarElement,
  ArcElement,
  PointElement,
  LineElement,
)

const component = defineAsyncComponent(async () => {
  const { Bar, Bubble, Doughnut, Line, Pie, PolarArea, Radar, Scatter } = await import(
    'vue-chartjs'
  )

  const components = {
    bar: Bar,
    line: Line,
    scatter: Scatter,
    bubble: Bubble,
    pie: Pie,
    doughnut: Doughnut,
    polarArea: PolarArea,
    radar: Radar,
  } as const

  return components[type]
})
</script>

<template>
  <!-- @vue-expect-error -->
  <Component :is="component" class="m-chart m-reset-css" :data :options :plugins :dataset-id-key :update-mode />
</template>
