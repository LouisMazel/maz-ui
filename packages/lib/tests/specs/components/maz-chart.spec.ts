import MazChart from '@package/components/MazChart.vue'
import { mount, shallowMount } from '@vue/test-utils'

const pieChart = {
  type: 'doughnut',
  data: {
    labels: [`Perfects - ${40}%`, `Bons - ${35}%`, `Mauvais - ${25}%`],
    datasets: [
      {
        backgroundColor: ['#fcb731', 'rgb(28 209 161)', 'rgb(255, 109, 106)'],
        data: [40, 35, 25],
      },
    ],
  },
}

describe('MazChart', () => {
  test('Should match wtesth the snapshot', () => {
    const wrapper = shallowMount(MazChart, {
      props: {
        data: pieChart.data,
        type: pieChart.type,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('renders correctly and emits event on click', async () => {
    const width = '300px'
    const height = '300px'
    const locale = 'fr-FR'

    const pieChart = {
      type: 'doughnut',
      data: {
        labels: [`Perfects - ${40}%`, `Bons - ${35}%`, `Mauvais - ${25}%`],
        datasets: [
          {
            backgroundColor: ['#fcb731', 'rgb(28 209 161)', 'rgb(255, 109, 106)'],
            data: [40, 35, 25],
          },
        ],
      },
    }

    const wrapper = mount(MazChart, {
      props: {
        data: pieChart.data,
        type: pieChart.type,
        width,
        height,
        locale,
      },
    })

    expect(wrapper.exists()).toBeTruthy()
    const canvas = wrapper.find('canvas')
    await canvas.trigger('click')
    expect(wrapper.emitted().click).toBeTruthy()
  })
})
