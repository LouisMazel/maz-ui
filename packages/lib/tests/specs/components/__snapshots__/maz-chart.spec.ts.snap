// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MazChart > Should match with  the snapshot 1`] = `""`;
